juju/juju
# Merge 2.9

Merge 2.9

#14088 Fix polarity of keep_env flag: keep_env == not teardown
#14083 [JUJU-1081] Fixes expose ipv6 issue in Openstack
#14092  [JUJU-1215] Fix timestamp when updating charmhub resources
#14097 [JUJU-1219] Make "juju clouds" show (client) clouds it can even if controller has an error
#14098 Update gopkg.in/yaml.v3
#14100  [JUJU-1221] updated CONTRIBUTING.md file
#14104  Rename Xcommand.go -> X.go
#14107  [JUJU-1220] Add tests to cover GetManifests not found case
#14108 [JUJU-1228] Remove patch QA files from tree
#14051  [JUJU-1125] Improve error message for local OCI image resource
#14052 [JUJU-1083] Make tests upgrade from 2.8
#14106 [JUJU-1226] Improve "no releases in channel" message when deploying
#14111  [JUJU-1232] Backport actions facade v7 support
#14112 [JUJU-1233] Backport backups facade v3 support

Most of the conflicts were from back porting facade changes and then forward merging. The resolution was to just accept the develop branch version.

```
# Conflicts:
#       api/client/action/client_test.go
#       api/client/action/run.go
#       api/client/action/run_test.go
#       api/client/action/types.go
#       api/facadeversions.go
#       apiserver/facades/client/action/action.go
#       apiserver/facades/client/action/register.go
#       apiserver/facades/client/action/run.go
#       apiserver/facades/client/action/run_test.go
#       apiserver/facades/client/backups/backups.go
#       apiserver/facades/client/backups/register.go
#       apiserver/facades/controller/charmrevisionupdater/updater.go
#       apiserver/facades/controller/charmrevisionupdater/updater_test.go
#       apiserver/facades/schema.json
#       charmstore/latest_test.go
#       cmd/juju/action/common.go
#       cmd/juju/action/package_test.go
#       cmd/juju/action/run.go
#       cmd/juju/action/runaction.go
#       cmd/juju/application/refresh.go
#       cmd/juju/commands/exec.go
#       cmd/juju/commands/exec_test.go
#       cmd/juju/metricsdebug/collectmetrics.go
#       cmd/juju/metricsdebug/collectmetrics_test.go
#       core/actions/actions.go
```

## QA steps
See PRs

[JUJU-1081]: https://warthogs.atlassian.net/browse/JUJU-1081?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[JUJU-1215]: https://warthogs.atlassian.net/browse/JUJU-1215?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[JUJU-1219]: https://warthogs.atlassian.net/browse/JUJU-1219?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[JUJU-1221]: https://warthogs.atlassian.net/browse/JUJU-1221?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[JUJU-1220]: https://warthogs.atlassian.net/browse/JUJU-1220?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[JUJU-1228]: https://warthogs.atlassian.net/browse/JUJU-1228?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[JUJU-1125]: https://warthogs.atlassian.net/browse/JUJU-1125?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[JUJU-1083]: https://warthogs.atlassian.net/browse/JUJU-1083?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[JUJU-1226]: https://warthogs.atlassian.net/browse/JUJU-1226?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[JUJU-1232]: https://warthogs.atlassian.net/browse/JUJU-1232?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[JUJU-1233]: https://warthogs.atlassian.net/browse/JUJU-1233?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
